source/exponential_backoff_retry_strategy.c

```diff
@@ -0,0 +1,351 @@
+    void *user_data,
+    uint64_t timeout_ms) {
+    (void)partition_id;
+    /* no resource contention here so no timeouts. */
+    (void)timeout_ms;
+
+    struct exponential_backoff_retry_token *backoff_retry_token =
+        aws_mem_calloc(retry_strategy->allocator, 1, sizeof(struct exponential_backoff_retry_token));
+
+    if (!backoff_retry_token) {
+        return AWS_OP_ERR;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_EXPONENTIAL_BACKOFF_RETRY_STRATEGY,
+        "id=%p: Initializing retry token %p",
+        (void *)retry_strategy,
+        (void *)&backoff_retry_token->base);
+
+    backoff_retry_token->base.allocator = retry_strategy->allocator;
+    backoff_retry_token->base.retry_strategy = retry_strategy;
+    aws_retry_strategy_acquire(retry_strategy);
+    backoff_retry_token->base.impl = backoff_retry_token;
+
+    struct exponential_backoff_strategy *exponential_backoff_strategy = retry_strategy->impl;
+    backoff_retry_token->bound_loop = aws_event_loop_group_get_next_loop(exponential_backoff_strategy->config.el_group);
+    backoff_retry_token->max_retries = exponential_backoff_strategy->config.max_retries;
+    backoff_retry_token->backofff_scale_factor_ns = aws_timestamp_convert(
+        exponential_backoff_strategy->config.backoff_scale_factor_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
+    backoff_retry_token->jitter_mode = exponential_backoff_strategy->config.jitter_mode;
+    backoff_retry_token->generate_random = exponential_backoff_strategy->config.generate_random;
+    aws_atomic_init_int(&backoff_retry_token->current_retry_count, 0);
+    aws_atomic_init_int(&backoff_retry_token->last_backoff, 0);
+
+    backoff_retry_token->thread_data.acquired_fn = on_acquired;
+    backoff_retry_token->thread_data.user_data = user_data;
+    AWS_FATAL_ASSERT(
+        !aws_mutex_init(&backoff_retry_token->thread_data.mutex) && "Retry strategy mutex initialization failed");
+
+    aws_task_init(
+        &backoff_retry_token->retry_task,
+        s_exponential_retry_task,
+        backoff_retry_token,
+        "aws_exponential_backoff_retry_task");
+    aws_event_loop_schedule_task_now(backoff_retry_token->bound_loop, &backoff_retry_token->retry_task);
+
+    return AWS_OP_SUCCESS;
+}
+
+static inline uint64_t s_random_in_range(uint64_t from, uint64_t to, struct exponential_backoff_retry_token *token) {
+    uint64_t max = aws_max_u64(from, to);
+    uint64_t min = aws_min_u64(from, to);
+
+    uint64_t diff = max - min;
+
+    if (!diff) {
+        return 0;
+    }
+
+    uint64_t random = token->generate_random();
+    return min + random % (diff);
+}
+
+typedef uint64_t(compute_backoff_fn)(struct exponential_backoff_retry_token *token);
+
+static uint64_t s_compute_no_jitter(struct exponential_backoff_retry_token *token) {
+    size_t retry_count = aws_atomic_load_int(&token->current_retry_count);
+    return aws_mul_u64_saturating((uint64_t)1 << retry_count, token->backofff_scale_factor_ns);
```

Comment:
`size_t retry_count = aws_min_size(aws_atomic_load_int(&token->current_retry_count), 63);`

Someone someday is going to remove the check at setup and everyone on the PR will just think "lol 63 what a random value"